source/connection_monitor.c

```diff
@@ -0,0 +1,200 @@
+
+static void s_process_statistics(
+    struct aws_crt_statistics_handler *handler,
+    struct aws_crt_statistics_sample_interval *interval,
+    struct aws_array_list *stats_list,
+    void *context) {
+
+    (void)interval;
+
+    struct aws_statistics_handler_http_connection_monitor_impl *impl = handler->impl;
+
+    uint64_t pending_read_interval_ns = 0;
+    uint64_t pending_write_interval_ns = 0;
+    uint64_t bytes_read = 0;
+    uint64_t bytes_written = 0;
+
+    /*
+     * Pull out the data needed to perform the throughput calculation
+     */
+    size_t stats_count = aws_array_list_length(stats_list);
+    for (size_t i = 0; i < stats_count; ++i) {
+        struct aws_crt_statistics_base *stats_base = NULL;
+        if (aws_array_list_get_at(stats_list, &stats_base, i)) {
+            continue;
+        }
+
+        switch (stats_base->category) {
+            case AWSCRT_STAT_CAT_SOCKET: {
+                struct aws_crt_statistics_socket *socket_stats = (struct aws_crt_statistics_socket *)stats_base;
+                bytes_read = socket_stats->bytes_read;
+                bytes_written = socket_stats->bytes_written;
+                break;
+            }
+
+            case AWSCRT_STAT_CAT_HTTP1: {
+                struct aws_crt_statistics_http1 *http1_stats = (struct aws_crt_statistics_http1 *)stats_base;
+                pending_read_interval_ns = http1_stats->pending_incoming_stream_ns;
+                pending_write_interval_ns = http1_stats->pending_outgoing_stream_ns;
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    struct aws_channel *channel = context;
+
+    if (impl->options.minimum_throughput_bytes_per_second == 0 ||
+        impl->options.minimum_throughput_failure_threshold_in_seconds == 0) {
+        return;
+    }
+
+    /*
+     * All early-out/negative execution paths reset the failure count to zero.  Keep a copy of the current count for the
+     * remaining path.
+     */
+    uint32_t current_failure_count = impl->consecutive_throughput_failures;
+    impl->consecutive_throughput_failures = 0;
+
+    uint64_t pending_io_time_ns = 0;
+    if (aws_add_u64_checked(pending_read_interval_ns, pending_write_interval_ns, &pending_io_time_ns)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_CHANNEL, "id=%p: io interval summation overflow", (void *)channel);
+        return;
+    }
+
+    if (pending_io_time_ns == 0) {
+        return;
+    }
+
+    uint64_t total_bytes = 0;
+    if (aws_add_u64_checked(bytes_read, bytes_written, &total_bytes)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_CHANNEL, "id=%p: io bytes summation overflow", (void *)channel);
+        return;
+    }
+
+    uint64_t bytes_per_second = 0;
+    if (aws_mul_u64_checked(total_bytes, (uint64_t)AWS_TIMESTAMP_NANOS, &bytes_per_second)) {
+        AWS_LOGF_INFO(AWS_LS_IO_CHANNEL, "id=%p: io throughput overflow calculation", (void *)channel);
+        return;
+    }
+
+    bytes_per_second /= pending_io_time_ns;
```

Comment:
maybe use `double` or `float` in these calculations so we don't have to worry about integer math chopping off remainders coming up with fishy numbers

Reply:
Done.